## Basic Bundle and File Operations

Here are some examples of basic operations with the HCA DSS: getting bundle and file metadata and contents. Here we'll illustrate this via direct calls to the API endpoints using [httpie](https://httpie.org).

First, install httpie so we can make some requests.

In [1]:
import sys
!{sys.executable} -m pip install httpie

Now, we're going to get the "manifest" of a bundle. This is metadata about a bundle and its contents. We'll construct a url based on the bundle's UUID and version.

In [2]:
bundle_uuid = "ead66505-a78b-44ee-81f6-418be859ab65"
bundle_version = "2018-12-06T043139.806469Z"

import urllib.parse
base_url = "https://dss.data.humancellatlas.org/v1/"
bundle_url = base_url + "bundles/" + bundle_uuid + '?' + urllib.parse.urlencode({"replica": "aws", "version": bundle_version})
print(f"Bundle url is: {bundle_url}")

Bundle url is: https://dss.data.humancellatlas.org/v1/bundles/ead66505-a78b-44ee-81f6-418be859ab65?replica=aws&version=2018-12-06T043139.806469Z


Now, we'll just GET that url.

In [3]:
!http GET "$bundle_url"

HTTP/1.1 200 OK
Access-Control-Allow-Headers: Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 8968
Content-Type: application/json
Date: Tue, 11 Dec 2018 23:03:29 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
X-Amzn-Trace-Id: Root=1-5c104241-50719294ff227f303b62fde2;Sampled=0
x-amz-apigw-id: Rw9KNEq6IAMFgZw=


{
    "bundle": {
        "creator_uid": 8008,
        "files": [
            {
                "content-type": "application/json; dcp-type=\"metadata/biomaterial\"",
                "crc32c": "ec0d3d14",
                "indexed": true,
                "name": "cell_suspension_0.json",
                "s3_etag": "fee4d8354468476052a1c0da67fb1f03",
                "sha1": "6f77abbdd0892a99e6fae5c920daaaf3018b24fd",
                "sha256": "5d41934e1af911f510be04506d41b59ebd91d1678f42df759819b223a898b1da",
                "size": 1356,
                "uuid": "

And there's the contents of that bundle along with some metadata. We can get metadata for a particular file by issuing a HEAD requests to a different url:

In [4]:
file_uuid = "9adf9f89-f546-4889-86ff-b430e3123c8b"
file_version = "2018-12-04T191256.554000Z"
file_url = base_url + "files/" + file_uuid + '?' + urllib.parse.urlencode({"replica": "aws", "version": file_version})
print(f"File url is: {file_url}")

File url is: https://dss.data.humancellatlas.org/v1/files/9adf9f89-f546-4889-86ff-b430e3123c8b?replica=aws&version=2018-12-04T191256.554000Z


In [5]:
!http HEAD "$file_url"

HTTP/1.1 200 OK
Access-Control-Allow-Headers: Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 0
Content-Type: text/html; charset=utf-8
Date: Tue, 11 Dec 2018 23:03:30 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
X-Amzn-Trace-Id: Root=1-5c104242-e86cdbe51a3f134449d6993d;Sampled=0
X-DSS-CONTENT-TYPE: application/json; dcp-type="metadata/process"
X-DSS-CRC32C: 3c8c3cc5
X-DSS-CREATOR-UID: 8008
X-DSS-S3-ETAG: 003989f2df2aa8f39723f3925e5fb851
X-DSS-SHA1: 9bff1801fb12b3302b7f7d7dd57a3a3cb6070914
X-DSS-SHA256: 7a967973684ffa8ac3fe5004177e2fc7d8f3a2c822bbede536dcd58d9af7bdb5
X-DSS-SIZE: 408
X-DSS-VERSION: 2018-12-04T191256.554000Z
x-amz-apigw-id: Rw9KbGWuIAMFR-Q=






This gives a response with metadata about the file in the header. If we want the actual contents of the file, we GET rather than HEAD that url:

In [6]:
!http GET "$file_url"

HTTP/1.1 302 Found
Access-Control-Allow-Headers: Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 1687
Content-Type: text/html; charset=utf-8
Date: Tue, 11 Dec 2018 23:03:32 GMT
Location: https://org-hca-dss-checkout-prod.s3.amazonaws.com/blobs/7a967973684ffa8ac3fe5004177e2fc7d8f3a2c822bbede536dcd58d9af7bdb5.9bff1801fb12b3302b7f7d7dd57a3a3cb6070914.003989f2df2aa8f39723f3925e5fb851.3c8c3cc5?AWSAccessKeyId=ASIARSZHKI4KOVDXHRWK&Signature=KW1YGr44t5ob%2F%2FgN%2BWawwd9WWIA%3D&x-amz-security-token=FQoGZXIvYXdzEIf%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaDGhmzBCT03o%2FJU6qGSLgAXSrzriN7DR2k04v4yE931hYaSm6niBbLcV9F8qvNahb8ab7FyjuoSfpUqYRvcq%2F%2Ft1bBBz9ORIkzBA6cCI%2BWFRr1qN6j7CdymmOhg9iS89icSROOtPDxV4xWAagzl3xUHpLT7OKuMlWiCosNXQaEU1UEVYPwGxVNdXKlDVIY5geh%2FbJ2sh%2B92uwiy16n0cURMY%2BCm8b%2F1SEnOrBZi5BklXmT14j%2BUzpToTVJfHVzt9PDbZnjQxaoV27CWTEple6J71XDNyVU%2BuoBHSbmqJqmlF%2BbD6wzMbFV42sWKc%2Bywt6KILNwOAF&Expires=1544

Okay, that didn't quite get us the file. It responded with a 302 and a "Location" in the header. If we want the file itself, we need to follow that redirect:

In [7]:
!http --follow GET "$file_url"

HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 408
Content-Type: application/json; dcp-type="metadata/process"
Date: Tue, 11 Dec 2018 23:03:35 GMT
ETag: "003989f2df2aa8f39723f3925e5fb851"
Last-Modified: Tue, 11 Dec 2018 23:00:17 GMT
Server: AmazonS3
x-amz-expiration: expiry-date="Wed, 19 Dec 2018 00:00:00 GMT", rule-id="dss_checkout_expiration"
x-amz-id-2: o2x3gUqcf/HnTpgUgcT9UeI8Wicag+L9R7XGfw3efs6Z4WAmiLp/mhOWe32hS78amKWF2IW+U4c=
x-amz-request-id: 0EF4D3BBBDB4B564


{
    "describedBy": "https://schema.humancellatlas.org/type/process/6.0.2/process",
    "process_core": {
        "process_id": "E18_20160930_Neurons_Sample_71_S068_L007_006"
    },
    "provenance": {
        "document_id": "9adf9f89-f546-4889-86ff-b430e3123c8b",
        "submission_date": "2018-12-04T18:59:53.218Z",
        "update_date": "2018-12-04T19:12:56.554Z"
    },
    "schema_type": "process"
}



This follows the redirect and gives us the contents of the (small) file in the response body.